CIP-array-builtin/README.md

```diff
@@ -0,0 +1,294 @@
+---
+CIP: ????
+Title: Plutus Core Builtin Type - `Array`
+Category: Plutus
+Status: Proposed
+Authors:
+    - Michael Peyton Jones <michael.peyton-jones@iohk.io>
+    - Ana Pantilie <ana.pantilie@iohk.io>
+Implementors: []
+Discussions:
+    - https://github.com/cardano-foundation/CIPs/pull/915
+Created: 2024-09-18
+License: CC-BY-4.0
+---
+
+## Abstract
+
+We propose an array builtin type for Plutus Core. This type will have constant-time lookup,
+which is a useful feature that is otherwise not possible to achieve. 
+
+## Motivation: why is this CIP necessary?
+
+The first part of [CPS-0013][1] outlines in great detail the motivation for introducing this
+new builtin type.
+
+To summarize, it is currently not possible to write a data structure with constant-time
+lookup in Plutus Core. We propose to solve this problem by introducing an array type into
+Plutus Core's builtin language, as it is the standard example of a data structure
+with this property, and it is a key component of many classical algorithms and data
+structures.
+
+Access to an array type would provide significant performance improvement opportunities to
+users of Plutus Core, since currently they must rely on suboptimal data structures such as
+lists for looking up elements inside a collection.
+
+## Specification
+
+We add the following builtin type: `Array` of kind `Type -> Type` representing
+a one-dimensional array type with indices of type `Integer`. Elements are indexed
+consecutively starting from `0`.
+
+***Note***: Here `Type` is the universe of all _builtin_ types, since we do not consider
+types formed out of applying builtin types to arbitrary types to be inhabited.
+
+The `Array` builtin type should be implemented with a fixed-size immutable array structure
+that has constant-time lookup.
+
+We add the following builtin functions:
+
+1. `indexArray` of type `forall a . Integer -> Array a -> a`.
+    - It returns the element at the given index in the array, or fails with an error if the
+    index is outside the bounds of the array.
+    - It uses constant time and constant memory.
+2. `lengthOfArray` of type `forall a . Array a -> Integer`
+    - It returns the length of the array.
+    - It uses constant time and constant memory.
+3. `listToArray` of type `forall a . List a -> Array a`
+    - It converts the argument builtin list into a builtin array.
+    - It uses linear time and linear memory.
+
+### Binary serialisation and deserialisation
```

Comment:
Does this cover [all](https://github.com/cardano-foundation/CIPs/blob/master/CIP-0035/README.md?plain=1#L225-L227) that CIP-035 requires of new types?

Reply:
Thank you for reminding me of this.

I asked @kwxm about this before I opened the PR. It's not very clear to us what the second point there refers to ("A precise description of the measure used for the size of a value of that type.").

If it's about measuring size for costing, I believe his opinion is that given that the two of the operations in the proposal are constant time/space and the third one can be costed in terms of the length of the list there's not much to be said here. Maybe just that the size of an array is equivalent to the size of the list of all its elements?

If that point refers to something else, maybe @michaelpj can clarify.